### ISSUE TITLE
Failed to serialize a 3D solid TopoDS_Shape

### ISSUE THREAD

**Issue by bimworld** [IfcOpenShell/IfcOpenShell#1126]
I get a none representation in python and a SIGSEGV in C++, trying to serialize the attached (.brep) shape. The brep analyzer does not throw an error. When trying to serialise the faces one by one, some are failing, others are not. I attached a face, that is failing.

[solid.zip](https://github.com/IfcOpenShell/IfcOpenShell/files/5591042/solid.zip)
[face.zip](https://github.com/IfcOpenShell/IfcOpenShell/files/5591172/face.zip)

**Comment by aothms** [comment]
In case serialize() fails it will return nullptr, so you need to check for this.

In this case serialize() fails because of `Geom_TrimmedCurve` being part of the topology which is currently not supported (and not recommended in OCCT. Better is to set the trim as the parameters of the edge vertices. Are you able to adapt your code like that?

**Comment by aothms** [comment]
Actually, come to think of it, it doesn't really matter whether a curve it trimmed geometrically, as in the current approach vertex is always specified simply by cartesian point. So I added a commit to just directly serialize the basis curve of the trimmed curve. Let me know if that helps.

**Comment by bimworld** [comment]
Thank you for looking into this. The shapes I want to serialize are processed by API functions of OCCT, like unifysamedomain, so I can't do anything to change the used geom classes, unfortunately.

Using your commit, I get now two different results:
1) advanced option is false

The function
```
int convert_to_ifc(const Handle_Geom_Curve& c, IfcSchema::IfcCurve*& curve, bool advanced)
``` 
is not called. The if statement of the function below, returns 0 and the Face is failing. Polygonal is set to false, because of crv->DynamicType() != STANDARD_TYPE(Geom_Line). 
```
    bool polygonal = true;
    for (TopExp_Explorer exp(wire, TopAbs_EDGE); exp.More(); exp.Next()) {
        double a, b;
        Handle_Geom_Curve crv = BRep_Tool::Curve(TopoDS::Edge(exp.Current()), a, b);
        if (crv.IsNull()) {
            continue;
        }
        if (crv->DynamicType() != STANDARD_TYPE(Geom_Line)) {
            polygonal = false;
            break;
        }
    }
    if (!polygonal && !advanced) {
        return 0; <------------------------------------------------------------
```

2) advanced option is true
It's running now. But I get advanced_faces only (as expected looking at the code). The problem is, that advanced_faces are not existent in Ifc2x3, so it would fail again.

Question 1)
As far as I think, the trimmed curves in the shape are all linear. Is there a way to not use advanced option and to check if the trimmed curve is linear and use the regular point to point connection, so the regular IfcFace is written?

Question 2)
Is it correct, that IFCFACETEDBREP and IFCADVANCEDBREP are used simultaniousely? Thats the result for advanced = true
```
#420935=IFCCARTESIANPOINT((33.1,13.7,0.));
#420936=IFCDIRECTION((0.,0.,1.));
#420937=IFCVECTOR(#420936,1.);
#420938=IFCLINE(#420935,#420937);
#420939=IFCEDGECURVE(#420932,#420934,#420938,.T.);
#420940=IFCORIENTEDEDGE(*,*,#420939,.T.);
#420941=IFCEDGELOOP((#420870,#420880,#420890,#420900,#420910,#420920,#420930,#420940));
#420942=IFCFACEOUTERBOUND(#420941,.T.);
#420943=IFCCARTESIANPOINT((33.,13.7,1.35));
#420944=IFCDIRECTION((0.,1.,0.));
#420945=IFCDIRECTION((0.,-0.,1.));
#420946=IFCAXIS2PLACEMENT3D(#420943,#420944,#420945);
#420947=IFCPLANE(#420946);
#420948=IFCADVANCEDFACE((#420942),#420947,.T.);
#420949=IFCCLOSEDSHELL((#420448,#420536,#420624,#420752,#420860,#420948));
#420950=IFCADVANCEDBREP(#420949);
#420951=IFCFACETEDBREP(#420949);
#420952=IFCSHAPEREPRESENTATION(#259892,'Body','Brep',(#420950,#420951));
#420953=IFCPRODUCTDEFINITIONSHAPE($,$,(#420952));
```

**Comment by aothms** [comment]
>  As far as I think, the trimmed curves in the shape are all linear. Is there a way to not use advanced option and to check if the trimmed curve is linear and use the regular point to point connection, so the regular IfcFace is written?

Good point, see 14193fb to check trimmed curve and also first degree nurbs curves.

>  Is it correct, that IFCFACETEDBREP and IFCADVANCEDBREP are used simultaniousely? Thats the result for advanced = true

Probably not. Also RepresentationType on the IfcShapeRepresentation should be 'AdvancedBrep' and not 'Brep' in that case. Maybe these are cases you can help with?

> Third problem is that after reading the new ifc file, faces are missing in the shape, maybe the reason is in question 2?

Let's evaluate when we have the other fixes in. Could also be related to tolerances for example or other factors.

**Comment by bimworld** [comment]
I didn't have a look at the code yetw, I just want to drop an overview of all possible cases:
[https://old.opencascade.com/doc/occt-7.0.0/refman/html/class_geom___geometry__inherit__graph.png](url)

**Comment by Moult** [comment]
This sounds fixed, can it be closed?

**Comment by BIMvoice** [comment]
Housekeeping: @Moult asked in 2023 whether this could be closed as fixed and there was no objection in the two years since. Seconding — safe to close.
